### ISSUE TITLE
ALT self-cast and other keys not working for the german keyboard

### ISSUE THREAD

**Issue by ineluki9** [loopuleasa/Dota2-TheCore-Config-Engine#103]
I am using the config files for the german keyboard.

In the untouched config ALT is remaped as follows:
dota_remap_alt_key "^"
This however does not work. But when I set my keyboard to english and remap alt to "`" (same key on the keyboard but other character) ALT is remaped correctly and self cast works.

Also the consol binding to "Ä" does not work.

I would really like to get this going for my german keyboard. Can you help me out?

cheers!

**Comment by loopuleasa** [comment]
Question 1, 2: The Reborn engine doesn't support all foreign key codes. This is volvo's part. The other contributor that made the german layout config is more familiar with these particularities.

Question 3: Unfortunately, we have no control over the item/spell labels from the cfg files. It's one downside of using the engine that made me sad.

**Comment by witchent** [comment]
I think I might be able to help with Question 1: Is your Keyboard set to latin1 with dead keys or without? So if you press your tilde key, does it write ^ or does it enable you to write the tilde on top of e.g. e?
Question 2: I have to try it out myself, which prob won't happen the next few days, but I will give you an update asap.

**Comment by ineluki9** [comment]
Thanks for the answers so far.

@loopuleasa: Just to clarify, volvo is another contributor you are working with? Regarding Question 3: I thought that you are not able to do from the cfg files. However why is it working with the items and not the abilities? Did you treat those two differently in the cfg files you made or is it game related?

@witchent: Q1: When I press ^ + e it writes: ê. When I press ^ just one time nothing happens. Pressing ^ 1 time plus space prints ^. My keyboard (Windows 10) is set to germany. Afaik this is the normal behaviour for the german standard keyboard, is it not? Do I have to change something here?

**Comment by Sembrani** [comment]
@ineluki9 volvo refers to Valve, the developer of DotA 2. It's a term used when Valve breaks things / does something wrong.

**Comment by witchent** [comment]
Q1: I'm not using windows, but in Linux I had to switch it so that it just writes ^ with a single press. Cause right now dota does not get your input, windows is waiting for another key to be pressed. You prob have to disable the dead key behavior, even though I have no idea how to do it in windows

**Comment by Sembrani** [comment]
@witchent @ineluki9 http://superuser.com/questions/346142/my-windows-keyboard-is-being-clever-with-the-quote-keys-how-can-i-stop-it

**Comment by ineluki9** [comment]
Thank you all for your quick support! 

@Sembrani, unfortunately the solution your link describes doesnt solve my problem since I dont want to uninstall german nor switch the system language to english each time I play dota.

Is anyone here actualy using the german cfg files or has tested them?
@loopuleasa, who is this german contributor you are speaking of?

**Comment by loopuleasa** [comment]
Yeah, they didn't offer proper key support for binds, which is a joke indeed.

As for the german layout, I am talking about this one https://github.com/loopuleasa/Dota2-TheCore-Config-Engine/tree/master/Dota%202%20Reborn%20Keyboard%20Setups/%5BGerman%20Keyboard%5D%20Loopuleasa's%20Super%20Compact%20QWERTG-DFXCV%20layout%20(ALT%2CSPACE%20mods)

**Comment by ineluki9** [comment]
Last question: Why is the remapping neccessary in the first place? In a earlier version the self cast was on space and everything worked fine. And why did it change?

**Comment by Sembrani** [comment]
It changed so that we could bind Space to normal cast, so there are 3 different "layers" of keybindings. It's largely a personal preference thing. If you would like, send me a layout of how you would like things, and I'll see if I can get a custom config to you soon.

**Comment by Sembrani** [comment]
@ineluki9 Another thing you could try is adding
```input_button_code_is_scan_code 1```
to the top of the autoexec.cfg file if you use the normal version.

Not sure if it will work since it was added in a recent update, though from my understanding, it should allow non-standard or non-QWERTY keyboards be bound using QWERTY keys.

**Comment by dodekeract** [comment]
@Sembrani I can confirm that at least the basic functionality of that command works with the German layout. I had no trouble binding using the English key names ever since I added that to [my personal autoexec](https://github.com/dodekeract/manta-config-engine/blob/master/presets/dodekeract.json).

**Comment by Sembrani** [comment]
So will it allow a german keyboard to use the english QWERTY binds?

**Comment by dodekeract** [comment]
@Sembrani Yes, at least for me. I tested ^, Ä, Y and Z. I assume the others work too. Didn't have trouble binding anything yet.

**Comment by ineluki9** [comment]
@Sembrani : what I meant is why you changed self-cast from space to alt. as far as i understand alt also has been a "layer" before (version 1: http://i.imgur.com/qXZ8bTs.png). and in that version you didnt had to remap alt and so that problem wasnt even present. personally I prefer the approach from version 1. pherhaps all the alt-remap issues would be gone by reverting it back to space. however that is just my opinion.

@Sembrani , @dodekeract the command is not working for me, put it at the top at the autoexec.cfg for the normal version 5.2. Neither TILDE works as a ALT replacement nore can I map actions to Ä.Ö and such. @dodekeract  if you should be feeling bored, you could give it a test with @loopuleasa's config :p

**Comment by dodekeract** [comment]
@ineluki9 did you try it with the English version? I can't test anything right now, since I'm on my notebook and my pc is not available atm.

**Comment by ineluki9** [comment]
Ok, it works (as far as I have testet it). Thanks @dodekeract and thanks @Sembrani  for the great work!

**Comment by kingalca** [comment]
How to disable dead keys for the ^-problem on Linux? @loopuleasa

**Comment by dodekeract** [comment]
I'm using a German keyboard layout on Kubuntu, it works fine for me when I use that "bind keys by keyboard position" setting and a normal English autoexec. You should definitely try that out.

**Comment by kingalca** [comment]
I have Ping now on ALT. I want ping to be on ^ on a german layout. How is that possible?

**Comment by loopuleasa** [comment]
The dota_remap_alt_key command in the main autoexec should be the one to do that, but I am unsure what codename that key has
